core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java

```diff
@@ -312,14 +335,16 @@ public void finishNow(@Nullable Throwable cause) {
         if (isFinishing()) {
             return;
         }
+        assert eventLoop != null;
+        if (!eventLoop.inEventLoop()) {
+            eventLoop.execute(() -> finishNow(cause));
+            return;
+        }
         if (isInitialized()) {
-            if (eventLoop.inEventLoop()) {
-                finishNow0(cause);
-            } else {
-                eventLoop.execute(() -> finishNow0(cause));
-            }
+            finishNow0(cause);
         } else {
-            addPendingTask(() -> finishNow0(cause));
+            start(noopCancellationTask);
+            finishNow0(cause);
```

Comment:
`ctx.cancel()` couldn't prevent the request headers from being written.
Would we abort the request in the following points if `ctx.isCanceled()` is true?
https://github.com/line/armeria/blob/56890684bc55994dbf866bcaa90d9ea86bf1aff2/core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java#L64-L64
https://github.com/line/armeria/blob/7911f5926e9efdef3d9c29e88e390ef1be560fc4/core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java#L218

Reply:
I'm not decided on how we want to set up the dependency between `log`, `HttpResponse`, and `ctx` yet but I think this is one possibility.

I think overall we want to
1. Hook callbacks so that there is a dependency between the above three states
2. Check whether a single state is completed or not

For now, I'm imagining that:
1. If `ctx` is cancelled, then the corresponding `HttpResponse` is cancelled
2. We check for `res.isDone` overall for early cancellation

But I will have to explore this further. For now, I think this PR doesn't change the previous behavior so it should probably be safe to merge

Reply:
> For now, I'm imagining that:

Sounds good.

The case where `ctx.cancel()` is called before a normal `CancellationTask` is registered can be considered a separate issue.